source/pele/pspin_spherical.h

```diff
@@ -0,0 +1,154 @@
+#ifndef _P_SPIN_SPHERICAL_H_
+#define _P_SPIN_SPHERICAL_H_
+
+#include "array.h"
+#include "base_potential.h"
+
+#include <iostream>
+#include <cassert>
+#include "pele/meta_pow.h"
+#include "pele/combination.h"
+
+namespace pele {
+
+template <size_t p>
+class MeanFieldPSpinSpherical : public BasePotential {
+protected:
+    pele::Array<double> m_interactions, m_spins;
+    pele::Array<size_t> m_indexes;
+    size_t m_N;
+    static const size_t m_p = p;
+    pele::combination_generator<size_t*> m_combination_generator;
+    pele::combination_generator<size_t*> m_combination_generator_grad;
+    size_t m_get_index(size_t* comb){
+        size_t idx = comb[0];
+        size_t d = m_N;
+        for (size_t i = 1; i < m_p; ++i){
+            idx += comb[i] * d;
+            d *= m_N;
+        }
+        return idx;
+    }
+public:
+    virtual ~MeanFieldPSpinSpherical() {}
+    //right now interactions_ is of size N**p which is not obtimal has it contains all the permutations of the interactions
+    //this is fast but costs a lot in terms of memory
+    MeanFieldPSpinSpherical(pele::Array<double> interactions_, size_t nspins_)
+            : m_interactions(interactions_.copy()),
+              m_spins(nspins_),
+              m_indexes(nspins_),
+              m_N(nspins_),
+              m_combination_generator(m_indexes.data(), m_indexes.data()+m_indexes.size(), m_p),
```

Comment:
fyi, Array can be used as an iterable, so you can use m_indices.begin() and m_indices.end() here